File: Part4/C#/SystemCalls.cs
using System;

class Program
{
    static void Main ()
    {
        // Note: For this program to function as expected, the user will have to correctly enter the requested values.

        // Boolean strings of true/false cannot be converted to a bool variable without conditions,
        // so we will discuss how that works in the branches video coming soon.
        // Some languages will not even allow for reading values of 0 or 1 from the terminal as
        // booleans so we will identify which languages this fails with, and revisit how to make 
        // this work in the branches video.

        string userInput;

        Console.Write ( "Type true and press enter." );
        userInput = Console.ReadLine ();
        bool parseSuccess = bool.TryParse ( userInput, out bool enteredBoolean );
        Console.WriteLine ( "The user entered the boolean {0} (parse succeeded: {1})", enteredBoolean, parseSuccess );

        Console.Write ( "Type false and press enter." );
        userInput = Console.ReadLine ();
        parseSuccess = bool.TryParse ( userInput, out enteredBoolean );
        Console.WriteLine ( "The user entered the boolean {0} (parse succeeded: {1})", enteredBoolean, parseSuccess );

        Console.Write ( "Type True and press enter." );
        userInput = Console.ReadLine ();
        parseSuccess = bool.TryParse ( userInput, out enteredBoolean );
        Console.WriteLine ( "The user entered the boolean {0} (parse succeeded: {1})", enteredBoolean, parseSuccess );

        Console.Write ( "Type FALSE and press enter." );
        userInput = Console.ReadLine ();
        parseSuccess = bool.TryParse ( userInput, out enteredBoolean );
        Console.WriteLine ( "The user entered the boolean {0} (parse succeeded: {1})", enteredBoolean, parseSuccess );

        Console.Write ( "Type 0 and press enter." );
        userInput = Console.ReadLine ();
        parseSuccess = bool.TryParse ( userInput, out enteredBoolean );
        Console.WriteLine ( "The user entered the boolean {0} (parse succeeded: {1})", enteredBoolean, parseSuccess );

        Console.Write ( "Type 1 and press enter." );
        userInput = Console.ReadLine ();
        parseSuccess = bool.TryParse ( userInput, out enteredBoolean );
        Console.WriteLine ( "The user entered the boolean {0} (parse succeeded: {1})", enteredBoolean, parseSuccess );

        Console.Write ( "Type 11 and press enter." );
        userInput = Console.ReadLine ();
        parseSuccess = bool.TryParse ( userInput, out enteredBoolean );
        Console.WriteLine ( "The user entered the boolean {0} (parse succeeded: {1})", enteredBoolean, parseSuccess );

        Console.Write ( "Type -1 and press enter." );
        userInput = Console.ReadLine ();
        parseSuccess = bool.TryParse ( userInput, out enteredBoolean );
        Console.WriteLine ( "The user entered the boolean {0} (parse succeeded: {1})", enteredBoolean, parseSuccess );

        Console.Write ( "Type 55 and press enter." );
        userInput = Console.ReadLine ();
        bool intParseSuccess = int.TryParse ( userInput, out int enteredInteger );
        Console.WriteLine ( "The user entered the integer {0} (parse succeeded: {1})", enteredInteger, intParseSuccess );

        Console.Write ( "Type 55.5 and press enter." );
        userInput = Console.ReadLine ();
        bool floatParseSuccess = float.TryParse ( userInput, out float enteredFloat );
        Console.WriteLine ( "The user entered the float {0} (parse succeeded: {1})", enteredFloat, floatParseSuccess );

        Console.Write ( "Type Hello World! and press enter." );
        userInput = Console.ReadLine ();
        Console.WriteLine ( "The user entered the string {0}", userInput );

        Console.Write ( "Type 123abc and press enter." );
        userInput = Console.ReadLine ();
        intParseSuccess = int.TryParse ( userInput, out enteredInteger );
        Console.WriteLine ( "The user entered the integer {0} (parse succeeded: {1})", enteredInteger, intParseSuccess );

        Console.Write ( "Type 123.45 and press enter." );
        userInput = Console.ReadLine ();
        intParseSuccess = int.TryParse ( userInput, out enteredInteger );
        Console.WriteLine ( "The user entered the integer {0} (parse succeeded: {1})", enteredInteger, intParseSuccess );

        Console.Write ( "Type abc123 and press enter." );
        userInput = Console.ReadLine ();
        intParseSuccess = int.TryParse ( userInput, out enteredInteger );
        Console.WriteLine ( "The user entered the integer {0} (parse succeeded: {1})", enteredInteger, intParseSuccess );

        Console.Write ( "Type  567 and press enter." );
        userInput = Console.ReadLine ();
        intParseSuccess = int.TryParse ( userInput, out enteredInteger );
        Console.WriteLine ( "The user entered the integer {0} (parse succeeded: {1})", enteredInteger, intParseSuccess );

        Console.Write ( "Type +567 and press enter." );
        userInput = Console.ReadLine ();
        intParseSuccess = int.TryParse ( userInput, out enteredInteger );
        Console.WriteLine ( "The user entered the integer {0} (parse succeeded: {1})", enteredInteger, intParseSuccess );

        Console.Write ( "Type -567 and press enter." );
        userInput = Console.ReadLine ();
        intParseSuccess = int.TryParse ( userInput, out enteredInteger );
        Console.WriteLine ( "The user entered the integer {0} (parse succeeded: {1})", enteredInteger, intParseSuccess );

        // C# has Math.Abs() that works with multiple types
        Console.WriteLine ( "Abs of -5 is {0}", Math.Abs ( -5 ) );
        Console.WriteLine ( "Abs of -5.5 is {0}", Math.Abs ( -5.5 ) );
        Console.WriteLine ( "Abs of a is {0}", Math.Abs ( (int)'a' ) );

        Console.WriteLine ( "Pow of 2^5 is {0}", Math.Pow ( 2, 5 ) );
        Console.WriteLine ( "Pow of 2.2^5.2 is {0}", Math.Pow ( 2.2, 5.2 ) );
        Console.WriteLine ( "Pow of a^b is {0}", Math.Pow ( (double)'a', (double)'b' ) );

        // Note trig functions are almost always in radians, not degrees
        Console.WriteLine ( "Sin of 90 is {0}", Math.Sin ( 90 ) );
        Console.WriteLine ( "Sin of pi/2 is {0}", Math.Sin ( Math.PI / 2 ) );

        Console.WriteLine ( "Cos of 180 is {0}", Math.Cos ( 180 ) );
        Console.WriteLine ( "Cos of pi is {0}", Math.Cos ( Math.PI ) );

        // Rounding type functions are very useful for explicit float to int conversions
        Console.WriteLine ( "Floor of 5.5 is {0}", Math.Floor ( 5.5 ) );
        Console.WriteLine ( "Floor of -5.5 is {0}", Math.Floor ( -5.5 ) );

        Console.WriteLine ( "Ceil of 5.5 is {0}", Math.Ceiling ( 5.5 ) );
        Console.WriteLine ( "Ceil of -5.5 is {0}", Math.Ceiling ( -5.5 ) );

        Console.WriteLine ( "Round of 5.5 is {0}", Math.Round ( 5.5 ) );
        Console.WriteLine ( "Round of -5.5 is {0}", Math.Round ( -5.5 ) );

        Console.WriteLine ( "Trunc of 5.5 is {0}", Math.Truncate ( 5.5 ) );
        Console.WriteLine ( "Trunc of -5.5 is {0}", Math.Truncate ( -5.5 ) );

        // This will NOT crash in C# (TryParse returns false for invalid input)
        Console.Write ( "Type Hello World! and press enter." );
        userInput = Console.ReadLine ();
        intParseSuccess = int.TryParse ( userInput, out enteredInteger );
        Console.WriteLine ( "The user entered the integer {0} (parse succeeded: {1})", enteredInteger, intParseSuccess );

        // This will NOT crash in C# (TryParse returns false for invalid input)
        Console.Write ( "Type abc123 and press enter." );
        userInput = Console.ReadLine ();
        intParseSuccess = int.TryParse ( userInput, out enteredInteger );
        Console.WriteLine ( "The user entered the integer {0} (parse succeeded: {1})", enteredInteger, intParseSuccess );
    }
}

File: Part4/Go/SystemCalls.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
)

func main () {
	// Note: For this program to function as expected, the user will have to correctly enter the requested values.

	// Boolean strings of true/false cannot be converted to a bool variable without conditions,
	// so we will discuss how that works in the branches video coming soon.
	// Some languages will not even allow for reading values of 0 or 1 from the terminal as
	// booleans so we will identify which languages this fails with, and revisit how to make 
	// this work in the branches video.

	scanner := bufio.NewScanner ( os.Stdin )

	// Note: Go's ParseBool only works with specific strings like "true", "false", "1", "0"
	// Other values like "11" or "-1" will cause errors
	fmt.Print ( "Type 0 and press enter.\n" )
	scanner.Scan ()
	userInput := scanner.Text ()
	enteredBoolean, err := strconv.ParseBool ( userInput )
	fmt.Printf ( "The user entered the boolean %t (error: %v)\n", enteredBoolean, err )

	fmt.Print ( "Type 1 and press enter.\n" )
	scanner.Scan ()
	userInput = scanner.Text ()
	enteredBoolean, err = strconv.ParseBool ( userInput )
	fmt.Printf ( "The user entered the boolean %t (error: %v)\n", enteredBoolean, err )

	fmt.Print ( "Type True and press enter.\n" )
	scanner.Scan ()
	userInput = scanner.Text ()
	enteredBoolean, err = strconv.ParseBool ( userInput )
	fmt.Printf ( "The user entered the boolean %t (error: %v)\n", enteredBoolean, err )

	fmt.Print ( "Type False and press enter.\n" )
	scanner.Scan ()
	userInput = scanner.Text ()
	enteredBoolean, err = strconv.ParseBool ( userInput )
	fmt.Printf ( "The user entered the boolean %t (error: %v)\n", enteredBoolean, err )

	fmt.Print ( "Type FALSE and press enter.\n" )
	scanner.Scan ()
	userInput = scanner.Text ()
	enteredBoolean, err = strconv.ParseBool ( userInput )
	fmt.Printf ( "The user entered the boolean %t (error: %v)\n", enteredBoolean, err )

	fmt.Print ( "Type 11 and press enter.\n" )
	scanner.Scan ()
	userInput = scanner.Text ()
	enteredBoolean, err = strconv.ParseBool ( userInput )
	fmt.Printf ( "The user entered the boolean %t (error: %v)\n", enteredBoolean, err )

	fmt.Print ( "Type -1 and press enter.\n" )
	scanner.Scan ()
	userInput = scanner.Text ()
	enteredBoolean, err = strconv.ParseBool ( userInput )
	fmt.Printf ( "The user entered the boolean %t (error: %v)\n", enteredBoolean, err )

	fmt.Print ( "Type 55 and press enter.\n" )
	scanner.Scan ()
	userInput = scanner.Text ()
	enteredInteger, err := strconv.Atoi ( userInput )
	fmt.Printf ( "The user entered the integer %d (error: %v)\n", enteredInteger, err )

	fmt.Print ( "Type 55.5 and press enter.\n" )
	scanner.Scan ()
	userInput = scanner.Text ()
	enteredFloat, err := strconv.ParseFloat ( userInput, 64 )
	fmt.Printf ( "The user entered the float %f (error: %v)\n", enteredFloat, err )

	fmt.Print ( "Type Hello World! and press enter.\n" )
	scanner.Scan ()
	userInput = scanner.Text ()
	fmt.Printf ( "The user entered the string %s\n", userInput )

	fmt.Print ( "Type 123abc and press enter.\n" )
	scanner.Scan ()
	userInput = scanner.Text ()
	enteredInteger, err = strconv.Atoi ( userInput )
	fmt.Printf ( "The user entered the integer %d (error: %v)\n", enteredInteger, err )

	fmt.Print ( "Type 123.45 and press enter.\n" )
	scanner.Scan ()
	userInput = scanner.Text ()
	enteredInteger, err = strconv.Atoi ( userInput )
	fmt.Printf ( "The user entered the integer %d (error: %v)\n", enteredInteger, err )

	fmt.Print ( "Type abc123 and press enter.\n" )
	scanner.Scan ()
	userInput = scanner.Text ()
	enteredInteger, err = strconv.Atoi ( userInput )
	fmt.Printf ( "The user entered the integer %d (error: %v)\n", enteredInteger, err )

	fmt.Print ( "Type  567 and press enter.\n" )
	scanner.Scan ()
	userInput = scanner.Text ()
	enteredInteger, err = strconv.Atoi ( userInput )
	fmt.Printf ( "The user entered the integer %d (error: %v)\n", enteredInteger, err )

	fmt.Print ( "Type +567 and press enter.\n" )
	scanner.Scan ()
	userInput = scanner.Text ()
	enteredInteger, err = strconv.Atoi ( userInput )
	fmt.Printf ( "The user entered the integer %d (error: %v)\n", enteredInteger, err )

	fmt.Print ( "Type -567 and press enter.\n" )
	scanner.Scan ()
	userInput = scanner.Text ()
	enteredInteger, err = strconv.Atoi ( userInput )
	fmt.Printf ( "The user entered the integer %d (error: %v)\n", enteredInteger, err )

	// Go has no built-in abs for integers, must write custom function or use math.Abs for floats
	fmt.Printf ( "Abs of -5 is %g\n", math.Abs ( -5 ) )
	fmt.Printf ( "Abs of -5.5 is %g\n", math.Abs ( -5.5 ) )
	fmt.Printf ( "Abs of a is %g\n", math.Abs ( float64 ( 'a' ) ) )

	fmt.Printf ( "Pow of 2^5 is %g\n", math.Pow ( 2, 5 ) )
	fmt.Printf ( "Pow of 2.2^5.2 is %g\n", math.Pow ( 2.2, 5.2 ) )
	fmt.Printf ( "Pow of a^b is %g\n", math.Pow ( float64 ( 'a' ), float64 ( 'b' ) ) )

	// Note trig functions are almost always in radians, not degrees
	fmt.Printf ( "Sin of 90 is %g\n", math.Sin ( 90 ) )
	fmt.Printf ( "Sin of pi/2 is %g\n", math.Sin ( 3.14159265358979323846/2 ) )

	fmt.Printf ( "Cos of 180 is %g\n", math.Cos ( 180 ) )
	fmt.Printf ( "Cos of pi is %g\n", math.Cos ( 3.14159265358979323846 ) )

	// Rounding type functions are very useful for explicit float to int conversions
	fmt.Printf ( "Floor of 5.5 is %g\n", math.Floor ( 5.5 ) )
	fmt.Printf ( "Floor of -5.5 is %g\n", math.Floor ( -5.5 ) )

	fmt.Printf ( "Ceil of 5.5 is %g\n", math.Ceil ( 5.5 ) )
	fmt.Printf ( "Ceil of -5.5 is %g\n", math.Ceil ( -5.5 ) )

	fmt.Printf ( "Round of 5.5 is %g\n", math.Round ( 5.5 ) )
	fmt.Printf ( "Round of -5.5 is %g\n", math.Round ( -5.5 ) )

	fmt.Printf ( "Trunc of 5.5 is %g\n", math.Trunc ( 5.5 ) )
	fmt.Printf ( "Trunc of -5.5 is %g\n", math.Trunc ( -5.5 ) )

	// This will NOT crash in Go (strconv.Atoi returns 0 and error for invalid input)
	fmt.Print ( "Type Hello World! and press enter.\n" )
	scanner.Scan ()
	userInput = scanner.Text ()
	enteredInteger, err = strconv.Atoi ( userInput )
	fmt.Printf ( "The user entered the integer %d (error: %v)\n", enteredInteger, err )

	// This will NOT crash in Go (strconv.Atoi returns 0 and error for invalid input)
	fmt.Print ( "Type abc123 and press enter.\n" )
	scanner.Scan ()
	userInput = scanner.Text ()
	enteredInteger, err = strconv.Atoi ( userInput )
	fmt.Printf ( "The user entered the integer %d (error: %v)\n", enteredInteger, err )
}
